presto-main/src/main/java/com/facebook/presto/operator/TableFinishOperator.java

```diff
@@ -342,4 +351,200 @@ public void close()
+        private final Map<LifespanStage, Map<Integer, LifespanStageState>> unfinishedRecoverableLifespanStageStates = new HashMap<>();
+        private final Map<LifespanStage, LifespanStageState> finishedRecoverableLifespanStages = new HashMap<>();
+
+        private final PartitionCommitter partitionCommitter;
+        private final JsonCodec<TableCommitContext> tableCommitContextCodec;
+
+        public LifespanStageStateTracker(PartitionCommitter partitionCommitter, JsonCodec<TableCommitContext> tableCommitContextCodec)
+        {
+            this.partitionCommitter = requireNonNull(partitionCommitter, "partitionCommitter is null");
+            this.tableCommitContextCodec = requireNonNull(tableCommitContextCodec, "tableCommitContextCodec is null");
+        }
+
+        public void update(Page page)
+        {
+            TableCommitContext tableCommitContext = getTableCommitContext(page);
+            LifespanStage lifespanStage = LifespanStage.fromTableCommitContext(tableCommitContext);
+            if (finishedRecoverableLifespanStages.containsKey(lifespanStage)) {
+                return;
+            }
+
+            switch (tableCommitContext.getCommitGranularity()) {
+                case TABLE:
+                    unrecoverableLifespanStageStates.putIfAbsent(lifespanStage, new LifespanStageState());
+                    unrecoverableLifespanStageStates.get(lifespanStage).update(page, false);
+                    break;
+                case PARTITION:
+                    unfinishedRecoverableLifespanStageStates.putIfAbsent(lifespanStage, new HashMap<>());
+                    Map<Integer, LifespanStageState> groupStageStatesPerTask = unfinishedRecoverableLifespanStageStates.get(lifespanStage);
+                    groupStageStatesPerTask.putIfAbsent(tableCommitContext.getTaskId(), new LifespanStageState());
+                    groupStageStatesPerTask.get(tableCommitContext.getTaskId()).update(page, true);
+
+                    if (isLastPageForGroupedLifespan(page)) {
+                        checkState(!finishedRecoverableLifespanStages.containsKey(lifespanStage), "GroupStage already finished");
+                        LifespanStageState lifespanStageState = groupStageStatesPerTask.get(tableCommitContext.getTaskId());
+                        finishedRecoverableLifespanStages.put(lifespanStage, lifespanStageState);
+                        unfinishedRecoverableLifespanStageStates.remove(lifespanStage);
+                        partitionCommitter.commitPartition(lifespanStage.getLifespan().getId(), lifespanStageState.getFragments());
+                    }
+                    break;
+                default:
+                    throw new IllegalStateException("Unrecognized commit granularity: " + tableCommitContext.getCommitGranularity());
+            }
+        }
+
+        public List<Page> getStatisticsPagesToProcess(Page page)
```

Comment:
nit: private. You can look at intellij warnings.

Reply:
It's actually package-private, but your point still holds :)